src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs

```diff
@@ -89,5 +94,56 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
             return lockFile;
         }
 
+        public async Task<PackagesLockFile> CreateNuGetLockFileAsync(IEnumerable<PackageReference> installedPackages, Func<PackageIdentity, string> getNupkgFilePath, CancellationToken token)
+        {
+            Debug.Assert(installedPackages != null);
+            Debug.Assert(getNupkgFilePath != null);
+
+            var lockFile = new PackagesLockFile();
+
+            foreach (var targetFramework in installedPackages.GroupBy(p => p.TargetFramework).OrderBy(g => g.Key))
+            {
+                var target = new PackagesLockFileTarget
+                {
+                    TargetFramework = targetFramework.Key,
+                    Dependencies = new List<LockFileDependency>()
+                };
+
+                foreach (var package in targetFramework.OrderBy(p => p.PackageIdentity))
+                {
+                    string hash;
+
+                    var file = getNupkgFilePath(package.PackageIdentity);
+                    if (file == null)
+                    {
+                        throw new Exception("couldn't find package");
+                    }
+
+                    using (var reader = new PackageArchiveReader(file))
+                    {
+                        hash = reader.GetContentHashForSignedPackage(token);
```

Comment:
I'd imagine by the time you call this API to generate lock file, packages have been installed so there'd be `.nupkg.metadata` file already present for each package. Please use that to get the content hash instead of regenerating that again.